### ISSUE TITLE
Hot reload on Shadow-CLJS bugs

### ISSUE THREAD

**Issue by mauricioszabo** [metosin/malli#695]
I decided to add Malli and instrumentation on my Shadow-CLJS project (Node.JS script). I'm not being able to make instrumentation + hot-reload work.

For example, I tried to add `(when js/goog.DEBUG (prn :instrumenting (malli.dev.cljs/start!)))` on my `core.cljs` file. Then, on another file, I added (as an example):

```clojure
(ns saturn.rings.services
  (:require [malli.core :as m]))

(defn example
  {:malli/schema [:=> [:cat string? int?] keyword?]}
  [a b]

  (prn :RAN)
  :ok)
```

And I also included this file on my `core.cljs` file. Now, when I save this file, it does indeed gets instrumented - but then things get weird. If I save this file, then `(example 10 20)` fails as expected. But if I change the code to:

```clojure
(defn example
  {:malli/schema [:=> [:cat int? int?] keyword?]}
  [a b]

  (prn :RAN)
  :nok)
```

And save *only once*, then the `(example "a" 20)` works, and it returns `:nok`, but the schema changed to `int?`, so this should fail... if I save a second time, then it works correctly. I also tried to run `(malli.dev.cljs/stop!)` before the start!, and the same result happens.

**Comment by dvingo** [comment]
Hi, there are some notes here: https://github.com/metosin/malli/blob/master/docs/clojurescript-function-instrumentation.md

because this works via macros you'll need to ensure all your code is loaded _and then_ apply the dev/start! (use shadow after-load hook)

**Comment by mauricioszabo** [comment]
Well, I _am using_ the after-load hook:

```clojure
(defn ^:dev/after-load main []
  (when js/goog.DEBUG (dev/start!))
  ...more init code...
```

**Comment by dvingo** [comment]
It's hard to diagnose without a sample repo. Most of the testing has been targeting the browser so there could be something different needed for node.js. A sample repository would help a lot to see what's going on.

**Comment by mauricioszabo** [comment]
Sure, I am making my experiments on Saturn, branch "services", but I generated a new one (`malli-bug-example`) to test this bug.

I am starting the instrument here: https://gitlab.com/clj-editors/saturn/-/blob/malli-bug-example/src/saturn/view.cljs#L173
And I am testing on this namespace: https://gitlab.com/clj-editors/saturn/-/blob/malli-bug-example/src/saturn/rings/services.cljs#L8

By the way, my first try I did not add this require: https://gitlab.com/clj-editors/saturn/-/blob/malli-bug-example/src/saturn/view.cljs#L16, and `(dev/start!)` did not capture anything to instrument, so maybe this also is something to look for?

**Comment by dvingo** [comment]
Thanks for the repo link. 

I think I see your issue. If you're uncommenting these lines and expecting instrumentation to work at this point it cannot:
https://gitlab.com/clj-editors/saturn/-/blob/malli-bug-example/src/saturn/rings/services.cljs#L12

If you read through the documentation I linked, the way this works is the function var is `(set!)` at the point where `dev/start!` is called. If the line is in the same file it cannot be checked yet.

using this example, things work fine:
```clojure
(defn ^:dev/after-load main []
  (when js/goog.DEBUG (prn :instrumenting (dev/start!)))
 
  (println "service: " (services/example 1 2))
  (println "service: " (services/example "1" 2))
```
![malli_instrument](https://user-images.githubusercontent.com/467827/164229694-62b1c2d6-c089-4220-b35a-057c3049ba03.png)

**Comment by mauricioszabo** [comment]
@dvingo no, it's not what I am doing. I am reloading things, and then evaluating the code (not uncommenting it). In fact, I can reproduce the error with the exact same code you sent me - I'm attaching a link to it (I added some try-catches, otherwise the code will not run. I also commented the tests so it's less clutter on the console)
![Malli Error](https://user-images.githubusercontent.com/138037/164234140-6289763f-9b4a-41d1-af3d-6876c14c60f8.gif)

**Comment by dvingo** [comment]
ok thanks for the explanation, this sort of report would have helped from the beginning, it wasn't clear what problems you were experiencing.

It appears there is some bug in the `dev/start!` macro. I will have to look further into this.

For now you can replace that call with these lines:

```clojure
	;; (require    [malli.instrument.cljs :as mi])
    (mi/unstrument!)
    (println "coll: " (dev/collect-all!))
    (mi/instrument!)
    (println "service: " (services/example 1 4)) 
```

This is what the `dev/start!` macro should be expanding to, but seems like there is some issue there.

**Comment by mauricioszabo** [comment]
Well, I did wrote:

> But if I change the code to:
> ...
> And save only once, then the (example "a" 20) works, and it returns :nok, but the schema changed to int?, so this should fail... if I save a second time, then it works correctly.

But no worries, the important thing is to be able to reproduce the error. Keep up the good work, I've been loving to use Malli :)

**Comment by dvingo** [comment]
Right, but there is a lot of nuance with the order of the code and the order of evaluation so without a sample project and the sequence of events it was hard to follow the exact problem. Thanks for supplying it!

FYI, as mentioned in the docs (https://github.com/metosin/malli/blob/master/docs/clojurescript-function-instrumentation.md#dev-setup) - making use of binaryage devtools will help a lot in inspecting the errors in the browser console. That's probably why you are having to catch the exceptions. I think devtools should work with electron.

**Comment by dvingo** [comment]
@mauricioszabo thanks again! Think I found a solution, will send a PR shortly

**Comment by dvingo** [comment]
@mauricioszabo can you try out the latest git sha? - this should work now

**Comment by mauricioszabo** [comment]
Sorry, I was trying to show another bug but I found that I can reproduce this same issue if I don't directly require the namespace where the instrumented functions appear.

I'll show another example.

**Comment by mauricioszabo** [comment]
Ok, it's the same bug, I'll not open a new one. Here's how I reproduce it:

I have two namespaces: `view.cljs`, that's my "core" namespace, and `services.cljs` that's where I'm instrumenting a function.

`services` is required from `events`
`events` is required from `view`

I'm not sure if it's because of this indirection, but `dev/start!` does not update the function I want to instrument _if I save_ `services` only. If I save `view`, then it will work sometimes, and it'll not work other tines. Here's a GIF showing the process:

![Bug](https://user-images.githubusercontent.com/138037/166323711-abb3d3a0-3706-42d8-a904-268db58fff21.gif)

Reproductible in: https://gitlab.com/clj-editors/saturn/-/commits/malli-bug-example2

**Comment by dvingo** [comment]
Oh man, that's annoying :) 

Instead of using `dev/start!` can you try just using:

`(malli.instrument.cljs/instrument! {:report (malli.dev.pretty/thrower)})`

when using `dev/start!` unstrument is called first which will set! back the original function. 
I am not sure that this is needed with hot reloading (since the functions will be overwritten anyway by the new code) and I'm wondering if that is what is causing this crazy behavior (where the set! from unstrument overwrites the newly hot-reloaded implementation).

-----

Another thing you can try, if you clone malli locally and use a `:local/root` dep, is to change this 
`defonce` to a `def` 
https://github.com/metosin/malli/blob/d6a8371fb50b86979d74ebc483c458d4d6181e3d/src/malli/instrument/cljs.cljs#L5
and then try seeing if `dev/start!` works then. By the same theory as above - I think this being a `defonce` is capturing the old function implementation when used during hot reloading.

**Comment by mauricioszabo** [comment]
Using your suggestions (`(malli.instrument.cljs/instrument! {:report (malli.dev.pretty/thrower)})`) gave me the same error, unfortunately :(.

I'll try to clone locally and make some changes to see if it works. By the way, the problem I'm having with Malli is _almost_ the same I had when I tried to use some combination on `spec` with `orchestra` in the past, so maybe a workaround could be to do the same thing that Prismatic Schema does (generate a `defn` that matches schema instead of first generating the function, _and then_ updating it to match the schema). Maybe on another _experimental_ namespace, WDYT?

**Comment by dvingo** [comment]
I have a fix in a branch for the scenario described above, but I don't think it's needed.

Instead you can attach this metadata to your entry namespace:

```clojure
(ns my-dev.entry
  {:dev/always true}
  (:require...))
```
and also still use only `instrument!` not `dev/start!`

Thomas Heller recommended it here:
https://clojureverse.org/t/problem-using-malli-clojurescript-instrumentation-and-shadow-cljs/8612/2

I don't think you need to go the preload route though, it can just be on the init namespace.

This will force the macro to get re-expanded whenever any files are changed, and thus should emit the most up-to-date schemas.

Try that out and see if it works for you - I have been testing it on a small app and it works well.

The benefit of this approach is that it supports the metadata style schemas also.

**Comment by dvingo** [comment]
After sleeping on it - I realized what you suggested makes sense - this is also the strategy used by guardrails. 
So why not have support for both styles??

Was able to get it working with minor changes to what's there too :D 

https://github.com/metosin/malli/compare/master...dvingo:experimental-cljs-defn?expand=1

**Comment by mauricioszabo** [comment]
That's wonderful, thanks a lot!

By the way - is it possible to also add a `>fn`, so we can have development-time instrumentation of anonymous functions?

**Comment by dvingo** [comment]
I'm curious what's the use case for that? you can just call  `(m/-instrument {:schema [:=> [:cat ...]} (fn [a b] c)` for such functionality

**Comment by mauricioszabo** [comment]
I'm thinking about callbacks, callbacks with higher order functions, and even "deep instrumentation" like, I can sign a function with:

```clojure
[:=> [:cat [:vector string?]
           [:=> [:cat string?] int?]]
     [:vector string?]]
```
and be able to check if the second argument indeed is receiving a string and returning an int

**Comment by Kauko** [comment]
Does #703 fix this?

**Comment by ikitommi** [comment]
about to make a release, long list of PRs for cljs-side, thanks @dvingo !!
